src/spartan_app.erl

```diff
@@ -124,11 +126,19 @@ load_json_config(FileBin) ->
 
 process_config_tuple({<<"upstream_resolvers">>, UpstreamResolvers}) ->
     UpstreamResolverIPs = lists:map(fun parse_ipv4_address/1, UpstreamResolvers),
-    ConfigValue = [{UpstreamResolverIP, 53} || UpstreamResolverIP <- UpstreamResolverIPs],
+    UpstreamResolverPorts = lists:map(fun parse_port_upstream_resolvers/1, UpstreamResolvers),
+    ConfigValue = [{UpstreamResolverIP, UpstreamResolverPort} || UpstreamResolverIP <- UpstreamResolverIPs, UpstreamResolverPort <-
+        UpstreamResolverPorts],
```

Comment:
I think this line is likely wrong. It will iterate all combinations of ports and pairs, for example:

```erlang
[{IP, Port} || IP <- ["127.0.0.1", "8.8.8.8"], Port <- [9000, 53]].
%% ⇒ [{"127.0.0.1",9000}, {"127.0.0.1",53}, {"8.8.8.8",9000}, {"8.8.8.8",53}]
```

We should fix the parsing and return the pair directly from the IP parsing function. I'd be happy to provide a patch.